docs/src/simulation_tips.md

```diff
@@ -0,0 +1,153 @@
+doesn't have to worry about details when setting up a simulation. However, there's just so much
+optimization that can be done in the source code. In order to take advantage of Julia's full speed
+the user has to be aware of some things when writing the simulation script. Furthermore, in case of
+more complex GPU runs, some details could sometimes prevent your simulation from running altogether.
+While Julia knowledge is obviously desirable here, a user that is unfamiliar with Julia can get away
+with efficient simulations by learning a few rules of thumb. It is also recommended that users go
+through Julia's [performance tips](https://docs.julialang.org/en/v1/manual/performance-tips/), which
+contains a more in-depth explanation of some of the aspects discussed here.
+
+
+## General (CPU/GPU) simulation tips
+
+### Avoid global variables whenever possible
+
+In general using a [global
+variable](https://docs.julialang.org/en/v1/manual/variables-and-scoping/#Global-Scope) (which can be
+loosely defined as a variable defined in the main script) inside functions slows down the code. One
+way to circumvent this is to always [use local variables or pass them as arguments to
+functions](https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-global-variables). This
+helps the compiler optimize the code.
+
+Another way around this is to [define global variables as constants whenever
+possible](https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-global-variables). One
+thing to keep in mind when doing this is that when a `const` is defined, its value can't be changed
+until you restart the Julia session. So this latter approach is good for production-ready code, but
+may be undesirable in the early stages of development while you still have to change the parameters
+of the simulation for exploration.
+
+
+### Consider inlining small functions
+
+Inlining is when the compiler [replaces a function call with the body of the function that is being
+called before compiling](https://en.wikipedia.org/wiki/Inline_expansion). The advantage of inlining
+(which is julia can be done with the [`@inline`
+macro](https://docs.julialang.org/en/v1/devdocs/meta/)) is that gets rid of the time spent calling the
+function. The Julia compiler automatically makes some calls as to what functions it should or
+shouldn't inline, but you can force a function to be inlined by including the macro `@inline` before
+its definition. This is more suited for small functions that are called often. Here's an example of
+an implementation of the Heaviside function that forces it to be inlined:
+
+```julia
+@inline heaviside(X) = ifelse(X < 0, zero(X), one(X))
+```
+
+In practice it's hard to say whether inlining a function will bring runtime benefits _with
+certainty_, since Julia already inlines some small functions automatically. However, it is generally
+a good idea to at least investigate this aspect in your code as the benefits can potentially be
+significant.
```

Comment:
It may be worth mentioning that `KernelAbstractions` "force inlines" all GPU code. I can't remember if CPU code is also force inlined. @vchuravy and @jakebolewski have give advice regarding the use of `@inline` in the past. Force-inlining suggests that we don't always need the annotation `@inline` but for some reason it still pervades all our code... ?

Reply:
I added a suggestion to mention KernelAbstractions.jl but it might not be useful to discuss inlining behavior to new users (many of whom are completely new to Julia).

My (probably unpopular) opinion is that I like to explicitly `@inline` functions to tell people reading my code that this is a performance-critical function I intend to have inlined whenever possible (even if it'll be force-inlined anyways). But this is probably a debate for another place.

Reply:
I think the docs on inlining in Julia are a bit scarce. There's a lot of spread out in discourse comments. For the time being I think this is the limit of my knowledge. But feel free to adapt the text if you think it needs more details.

Reply:
I think its fine to recommend `@inline` for style / clarity. But we shouldn't mislead people in the docs by saying that `@inline` is necessary for performance if it has no effect.

Reply:
@glwagner are you suggesting that `@inline` doesn't impact performance?

Reply:
I don't think it does for GPU, but it might for CPU...

Reply:
Could be fun to illustrate these points with code snippets that do some benchmarking that users can use for themselves (some day).****